# Deep Fool

DeepFool is an adversarial attack algorithm introduced by Moosavi-Dezfooli et al. (2016) that generates minimal perturbations to cause neural network misclassification. Unlike other adversarial attacks that may produce unnecessarily large perturbations, DeepFool takes a geometric approach by iteratively computing the smallest perturbation required to push an input sample across the nearest decision boundary. This makes it particularly useful for measuring the robustness of neural networks, as it provides a lower bound on the perturbation magnitude needed to fool the model.

The algorithm treats the classification problem geometrically, where each class region is separated by decision boundaries (hyperplanes in the linearized case). Starting from a correctly classified input, DeepFool iteratively moves the sample toward the closest decision boundary by computing minimal perturbations in the direction perpendicular to each separating hyperplane. The attack terminates when the perturbed input crosses into a different class region, resulting in misclassification.

## Implementation

Our implementation of DeepFool follows the algorithm presented in the original paper, with modifications to handle both CNN and RNN architectures for audio classification. At each iteration, the algorithm computes the gradients (Jacobian) of all class logits with respect to the input. For each class *k* different from the current predicted class, it calculates the hyperplane defined by the difference between class *k* and the true class. The minimal perturbation to reach each hyperplane is computed as **r = |f_l| / ||w_l||² · w_l**, where *f_l* is the difference in logit values and *w_l* is the difference in gradients between the two classes. The algorithm selects the hyperplane with the smallest distance and updates the total perturbation accordingly. This process repeats until the model's prediction changes or a maximum number of iterations is reached. A small overshoot parameter (typically 2%) is applied to the final perturbation to ensure the perturbed sample crosses the decision boundary reliably. For RNN models, we apply a specific workaround that temporarily enables training mode during gradient computation to satisfy cuDNN's requirements for LSTM backward passes, while keeping batch normalization and dropout layers in evaluation mode throughout the attack.

In [1]:
import torch
import torch.nn as nn

# Helper functions for RNN gradient management
def enable_rnn_grad(m):
    """Enable train mode for RNN layers to allow backward pass."""
    if not isinstance(m, (nn.BatchNorm2d, nn.Dropout, nn.Dropout2d)):
        m.train()

def disable_rnn_grad(m):
    """Set the entire model back to evaluation mode."""
    m.eval()

def is_rnn_model(model):
    """Check if model contains RNN layers (LSTM, GRU)."""
    for module in model.modules():
        if isinstance(module, (nn.LSTM, nn.GRU, nn.RNN)):
            return True
    return False

def compute_jacobian_and_values(model, x, lengths):
    """
    Computes the function values (logits) and the Jacobian (gradients)
    for all classes w.r.t the input x.
    """
    # Compute logits
    f_x = model(x.unsqueeze(0), lengths)
    f_x = f_x.squeeze(0)

    num_classes = f_x.numel()
    jacobian_dict = {}

    for k in range(num_classes):
        model.zero_grad()
        f_k = f_x[k]
        gradient_k = torch.autograd.grad(f_k, x, retain_graph=True)[0]
        jacobian_dict[k] = gradient_k
        
    return f_x, jacobian_dict

def deep_fool_attack(model, x_original, lengths, max_iter=50, overshoot=0.02, num_classes=10, device='cpu'):
    """
    Performs the DeepFool adversarial attack.
    Works with both CNN and RNN models.
    """
    # Check if model is RNN-based
    use_rnn_workaround = is_rnn_model(model)
    
    if use_rnn_workaround:
        model.apply(disable_rnn_grad)
    else:
        model.eval()
    
    x_i = x_original.clone().detach().to(device)
    r_total = torch.zeros_like(x_i)
    loop_iter = 0
    
    # Initial prediction
    with torch.no_grad():
        initial_logits = model(x_i.unsqueeze(0), lengths).squeeze(0)
    
    initial_label = initial_logits.argmax().item()
    current_label = initial_label
    
    x_i.requires_grad_(True)
    
    while current_label == initial_label and loop_iter < max_iter:
        
        # Enable train mode for RNN backward pass if needed
        if use_rnn_workaround:
            model.apply(enable_rnn_grad)
        
        f_x_i, w_k_dict = compute_jacobian_and_values(model, x_i, lengths)
        
        if use_rnn_workaround:
            model.apply(disable_rnn_grad)
        
        f_initial = f_x_i[initial_label]
        w_initial = w_k_dict[initial_label]
        
        # Find the closest decision boundary
        min_dist = float('inf')
        min_r_i = None
        
        for k in range(num_classes):
            if k == initial_label:
                continue
            
            f_k = f_x_i[k]
            w_k = w_k_dict[k]
            
            # Hyperplane parameters
            w_l = w_k - w_initial
            f_l = f_k - f_initial
            
            w_l_norm_sq = torch.sum(w_l ** 2)
            
            if w_l_norm_sq < 1e-8:
                continue
            
            # Calculate perturbation to reach this hyperplane
            # r_i = |f_l| / ||w_l||^2 * w_l
            r_i = torch.abs(f_l) / w_l_norm_sq * w_l
            
            # Distance to this hyperplane
            dist = torch.abs(f_l) / torch.sqrt(w_l_norm_sq)
            
            if dist < min_dist:
                min_dist = dist
                min_r_i = r_i
                
        if min_r_i is None:
            break
        
        # Accumulate perturbation
        r_total = r_total + min_r_i.detach()
        
        # Update x_i with accumulated perturbation
        x_i = x_original.clone().detach().to(device) + r_total
        x_i.requires_grad_(True)
        
        # Check new prediction
        with torch.no_grad():
            current_logits = model(x_i.unsqueeze(0), lengths).squeeze(0)
            current_label = current_logits.argmax().item()
        
        loop_iter += 1
    
    # Apply overshoot
    r_total = (1 + overshoot) * r_total
    
    # Final verification
    with torch.no_grad():
        final_label = model((x_original + r_total).unsqueeze(0), lengths).squeeze(0).argmax().item()
    
    return r_total, loop_iter, final_label

In [18]:
import os
import numpy as np
import pandas as pd

from datetime import datetime
from torch.utils.data import DataLoader

from models.cnn import AudioCNN, LazyAudioCNNDataset, collate_fn_cnn
from models.rnn import AudioRNN, LazyAudioRNNDataset, collate_fn_rnn
from utils import load_fold_paths, read_config

In [19]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
config = read_config('../config/rnn.json')

base_dir = '../data_cache'
folders = [f for f in os.listdir(base_dir) if os.path.isdir(os.path.join(base_dir, f))]

folder_dates = []
for folder in folders:
    try:
        folder_date = datetime.strptime(folder, '%m%d_%H%M%S')
        folder_dates.append((folder_date, folder))
    except ValueError:
        print(f"Folder name '{folder}' does not match expected format.")

if folder_dates:
    most_recent_folder = max(folder_dates, key=lambda x: x[0])[1]
    data_cache_dir = os.path.join(base_dir, most_recent_folder)
    print(f"Most recent folder: {data_cache_dir}")
else:
    print("No valid folders found.")

SAMPLES_PER_FOLD = 100

FOLDS_TO_TEST = [1, 3, 5]

RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
torch.manual_seed(RANDOM_SEED)

rnn_deepfool_results = []
_, _, test_paths, test_labels = load_fold_paths(data_cache_dir=data_cache_dir)

for fold in range(1, 11):
    if fold not in FOLDS_TO_TEST:
        print(f"\n=== Skipping Fold {fold} ===")
        continue
    
    print(f"\n=== RNN Fold {fold} ===")

    ckpt = torch.load(f"../saved_cv/rnn/1130_175220/fold_5/best_model.pth", map_location=device)

    rnn_model = AudioRNN(config=ckpt["config"]).to(device)
    rnn_model.load_state_dict(ckpt["model_state_dict"])
    rnn_model.eval()

    X_test = test_paths[fold]
    y_test = test_labels[fold]

    test_dataset = LazyAudioRNNDataset(X_test, y_test)
    test_loader = DataLoader(test_dataset, batch_size=config['batch_size'], 
                            shuffle=False, num_workers=0, collate_fn=collate_fn_rnn)

    all_correct_samples = []
    
    for batch_mels, batch_lengths, batch_labels in test_loader:
        for mel, lengths, label in zip(batch_mels, batch_lengths, batch_labels):
            mel = mel.to(device)
            lengths = lengths.unsqueeze(0).to(device)
            label_scalar = label.item()
    
            with torch.no_grad():
                pred = rnn_model(mel.unsqueeze(0), lengths).argmax().item()
    
            if pred == label_scalar:
                all_correct_samples.append((mel, lengths, label_scalar))
    
    print(f"Total correctly classified samples: {len(all_correct_samples)}")
    
    n_samples = min(SAMPLES_PER_FOLD, len(all_correct_samples))
    
    sampled_indices = np.random.choice(len(all_correct_samples), 
                                      size=n_samples, 
                                      replace=False)
    
    print(f"Testing DeepFool on {n_samples} samples ({100*n_samples/len(all_correct_samples):.1f}%)")
    
    total = 0
    fooled = 0
    perturb_norms = []

    # ----------------------- DeepFool Loop on Sampled Data -----------------------
    for idx in sampled_indices:
        mel, lengths, label_scalar = all_correct_samples[idx]
        
        mel_attack = mel.clone().detach().requires_grad_(True)
        
        r, _, final_label = deep_fool_attack(
            rnn_model,
            mel_attack,
            lengths=lengths,
            num_classes=10,
            device=device
        )

        perturb_norms.append(r.norm().item())
        fooled += (final_label != label_scalar)
        total += 1

    rnn_deepfool_results.append({
        "fold": fold,
        "samples_tested": total,
        "total_correct": len(all_correct_samples),
        "fool_rate": fooled / max(1, total),
        "mean_perturb_norm": float(np.mean(perturb_norms)) if perturb_norms else None
    })

df_rnn = pd.DataFrame(rnn_deepfool_results)
print("\n" + "="*60)
print("DEEPFOOL RESULTS (SAMPLED DATA)")
print("="*60)
display(df_rnn)

print("\n==== Averages Across Tested Folds ====")
print(f"Average Fool Rate: {df_rnn['fool_rate'].mean():.4f}")
print(f"Average Perturbation Norm: {df_rnn['mean_perturb_norm'].mean():.4f}")
print(f"Total Samples Tested: {df_rnn['samples_tested'].sum()}")

Most recent folder: ../data_cache/1128_230416

=== RNN Fold 1 ===
Total correctly classified samples: 684
Testing DeepFool on 100 samples (14.6%)

=== Skipping Fold 2 ===

=== RNN Fold 3 ===
Total correctly classified samples: 646
Testing DeepFool on 100 samples (15.5%)

=== Skipping Fold 4 ===

=== RNN Fold 5 ===
Total correctly classified samples: 585
Testing DeepFool on 100 samples (17.1%)

=== Skipping Fold 6 ===

=== Skipping Fold 7 ===

=== Skipping Fold 8 ===

=== Skipping Fold 9 ===

=== Skipping Fold 10 ===

DEEPFOOL RESULTS (SAMPLED DATA)


,fold,samples_tested,total_correct,fool_rate,mean_perturb_norm
0,1,100,684,1.0,9.807635
1,3,100,646,1.0,41.557846
2,5,100,585,1.0,71.825453



==== Averages Across Tested Folds ====
Average Fool Rate: 1.0000
Average Perturbation Norm: 41.0636
Total Samples Tested: 300


In [21]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
config = read_config('../config/cnn.json')

base_dir = '../data_cache'
folders = [f for f in os.listdir(base_dir) if os.path.isdir(os.path.join(base_dir, f))]

folder_dates = []
for folder in folders:
    try:
        folder_date = datetime.strptime(folder, '%m%d_%H%M%S')
        folder_dates.append((folder_date, folder))
    except ValueError:
        print(f"Folder name '{folder}' does not match expected format.")

if folder_dates:
    most_recent_folder = max(folder_dates, key=lambda x: x[0])[1]
    data_cache_dir = os.path.join(base_dir, most_recent_folder)
    print(f"Most recent folder: {data_cache_dir}")
else:
    print("No valid folders found.")

SAMPLES_PER_FOLD = 100

FOLDS_TO_TEST = [1, 3, 5]

RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
torch.manual_seed(RANDOM_SEED)

cnn_deepfool_results = []
_, _, test_paths, test_labels = load_fold_paths(data_cache_dir=data_cache_dir)

for fold in range(1, 11):
    if fold not in FOLDS_TO_TEST:
        print(f"\n=== Skipping Fold {fold} ===")
        continue
    
    print(f"\n=== CNN Fold {fold} ===")

    ckpt = torch.load(f"../saved_cv/cnn/1130_211130/fold_5/best_model.pth", map_location=device)

    cnn_model = AudioCNN(config=ckpt["config"]).to(device)
    cnn_model.load_state_dict(ckpt["model_state_dict"])
    cnn_model.eval()

    X_test = test_paths[fold]
    y_test = test_labels[fold]

    test_dataset = LazyAudioCNNDataset(X_test, y_test)
    test_loader = DataLoader(test_dataset, batch_size=config['batch_size'], 
                            shuffle=False, num_workers=0, collate_fn=collate_fn_cnn)

    all_correct_samples = []
    
    for batch_mels, batch_lengths, batch_labels in test_loader:
        for mel, lengths, label in zip(batch_mels, batch_lengths, batch_labels):
            mel = mel.to(device)
            lengths = lengths.unsqueeze(0).to(device)
            label_scalar = label.item()
    
            with torch.no_grad():
                pred = cnn_model(mel.unsqueeze(0), lengths).argmax().item()
    
            if pred == label_scalar:
                all_correct_samples.append((mel, lengths, label_scalar))
    
    print(f"Total correctly classified samples: {len(all_correct_samples)}")
    
    n_samples = min(SAMPLES_PER_FOLD, len(all_correct_samples))
    
    sampled_indices = np.random.choice(len(all_correct_samples), 
                                      size=n_samples, 
                                      replace=False)
    
    print(f"Testing DeepFool on {n_samples} samples ({100*n_samples/len(all_correct_samples):.1f}%)")
    
    total = 0
    fooled = 0
    perturb_norms = []

    # ----------------------- DeepFool Loop on Sampled Data -----------------------
    for idx in sampled_indices:
        mel, lengths, label_scalar = all_correct_samples[idx]
        
        mel_attack = mel.clone().detach().requires_grad_(True)
        
        r, _, final_label = deep_fool_attack(
            cnn_model,
            mel_attack,
            lengths=lengths,
            num_classes=10,
            device=device
        )

        perturb_norms.append(r.norm().item())
        fooled += (final_label != label_scalar)
        total += 1

    cnn_deepfool_results.append({
        "fold": fold,
        "samples_tested": total,
        "total_correct": len(all_correct_samples),
        "fool_rate": fooled / max(1, total),
        "mean_perturb_norm": float(np.mean(perturb_norms)) if perturb_norms else None
    })

df_cnn = pd.DataFrame(cnn_deepfool_results)
print("\n" + "="*60)
print("DEEPFOOL RESULTS (SAMPLED DATA)")
print("="*60)
display(df_cnn)

print("\n==== Averages Across Tested Folds ====")
print(f"Average Fool Rate: {df_cnn['fool_rate'].mean():.4f}")
print(f"Average Perturbation Norm: {df_cnn['mean_perturb_norm'].mean():.4f}")
print(f"Total Samples Tested: {df_cnn['samples_tested'].sum()}")

Most recent folder: ../data_cache/1128_230416

=== CNN Fold 1 ===
Total correctly classified samples: 738
Testing DeepFool on 100 samples (13.6%)

=== Skipping Fold 2 ===

=== CNN Fold 3 ===
Total correctly classified samples: 638
Testing DeepFool on 100 samples (15.7%)

=== Skipping Fold 4 ===

=== CNN Fold 5 ===
Total correctly classified samples: 669
Testing DeepFool on 100 samples (14.9%)

=== Skipping Fold 6 ===

=== Skipping Fold 7 ===

=== Skipping Fold 8 ===

=== Skipping Fold 9 ===

=== Skipping Fold 10 ===

DEEPFOOL RESULTS (SAMPLED DATA)


,fold,samples_tested,total_correct,fool_rate,mean_perturb_norm
0,1,100,738,1.0,3.475612
1,3,100,638,1.0,3.799358
2,5,100,669,1.0,4.513592



==== Averages Across Tested Folds ====
Average Fool Rate: 1.0000
Average Perturbation Norm: 3.9295
Total Samples Tested: 300


# Conclusion

## DeepFool Attack Results

Our DeepFool adversarial attack analysis revealed significant vulnerabilities in both the CNN and RNN audio classification models. Testing on sampled data across multiple folds (folds 1, 3, and 5, with 100 samples per fold), we achieved a 100% fooling rate for both architectures, demonstrating that adversarial perturbations can successfully cause misclassification on all correctly predicted samples tested.

The key finding lies in the stark difference in robustness between the two architectures. The **RNN model exhibited substantially greater robustness** with an average perturbation norm of approximately **41.06**, while the **CNN required only minimal perturbations** averaging **3.93** to achieve misclassification. This represents roughly a **10-fold difference** in adversarial robustness, with RNNs demonstrating superior resistance to adversarial attacks in the audio classification domain.

The 100% success rate across both models indicates that while adversarial attacks can eventually fool both architectures, the magnitude of perturbation required differs dramatically. The CNN's lower perturbation threshold suggests that its decision boundaries are closer to correctly classified samples, making it more vulnerable to small adversarial perturbations. In contrast, the RNN's significantly higher perturbation requirement indicates more separated decision boundaries and greater inherent robustness.

## Implications and Future Work

These results highlight the importance of adversarial robustness evaluation in audio classification systems, particularly for security-critical applications. The dramatic difference in perturbation magnitudes between CNN and RNN architectures suggests that **recurrent approaches may be inherently more robust** to adversarial attacks in the audio domain, possibly due to their temporal modeling capabilities and sequential processing of audio features that may create more complex decision boundaries.

Future work should explore adversarial training techniques to improve model robustness, particularly for the CNN architecture which showed greater vulnerability. Additional research should investigate the relationship between model architecture and adversarial vulnerability in audio tasks, and examine whether ensemble methods combining CNN and RNN architectures could leverage the RNN's robustness while maintaining the CNN's computational efficiency.